ocaml/database/slave_backup_master_connection.ml

```diff
@@ -0,0 +1,262 @@
+(*
+ * Copyright (C) 2006-2009 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+(* Manages persistent connection from slave->master over which the db is accessed.
+
+   The state of this connection is used by the slave to determine whether it can see its
+   master or not. After the slave has not been able to see the master for a while, xapi
+   restarts in emergency mode.
+*)
+
+open Xapi_stdext_threads.Threadext
+
+type db_record = (string * string) list * (string * (string list)) list
+module D = Debug.Make(struct let name = "slave_backup_master_connection" end)
+open D
+
+let my_connection : Stunnel.t option ref = ref None
+
+exception Uninitialised
+
+let is_slave : (unit -> bool) ref = ref (fun () -> error "is_slave called without having been set. This is a fatal error."; raise Uninitialised)
+let get_master_address = ref (fun () -> error "get_master_address called without having been set. This is a fatal error"; raise Uninitialised)
+let master_rpc_path = ref "<invalid>"
+
+exception Cannot_connect_to_master
+
+(* kill the stunnel underlying the connection.. When the master dies
+   then read/writes to the connection block for ages waiting for the
+   TCP timeout. By killing the stunnel, we can get these calls to
+   unblock. (We do not clean up after killing stunnel, it is still the
+   duty of the person who initiated the stunnel connection to call
+   Stunnel.disconnect which closes the relevant fds and does wait_pid on
+   the (now dead!) stunnel process.
+*)
+let force_connection_reset () =
```

Comment:
This file is now mostly a copy of master_connection.ml, which is not what we want. We need to make master_connection more generic where needed, while being careful about all the parts of xapi that interact with it. E.g., this `force_connection_reset` function must probably always be called on both connections.

Since `Master_connection` has so many `ref` values on the top level, we probably need to turn most of it into a functor, while providing a common high-level interface where needed.

Reply:
For reference, the "interesting" interactions with the master connection are things like management IP addresses changes and pool master changes/failover.